File: web/src/sim/store.ts
/** Bulk commodities tracked by mass. Cabin atmosphere is NOT a store (see Habitat). */
export type Resource =
  | 'o2'
  | 'n2'
  | 'co2'
  | 'h2'
  | 'ch4'
  | 'lox'
  | 'waterPotable'
  | 'waterWaste'
  | 'waterFeedstock'
  | 'food'
  | 'biomass'
  | 'regolith'
  | 'sparesMass';

export const resourceDisplayName: Record<Resource, string> = {
  o2: 'Oxygen reserve',
  n2: 'Nitrogen',
  co2: 'CO₂ buffer',
  h2: 'Hydrogen',
  ch4: 'Methane (propellant)',
  lox: 'LOX (propellant)',
  waterPotable: 'Potable water',
  waterWaste: 'Wastewater',
  waterFeedstock: 'Raw water (ISRU)',
  food: 'Food',
  biomass: 'Biomass',
  regolith: 'Regolith',
  sparesMass: 'Spares',
};

/**
 * A mass buffer with capacity. All withdrawals/deposits are clamped and ledgered so the
 * engine can audit conservation exactly: deposited − withdrawn === amount.
 */
export class Store {
  readonly id: string;
  readonly resource: Resource;
  displayName: string;

  capacityKg: number;
  amountKg: number;

  totalDepositedKg: number;
  totalWithdrawnKg = 0;
  totalVentedKg = 0; // deposits rejected for lack of capacity (never entered the store)

  constructor(id: string, resource: Resource, capacityKg: number, initialKg = 0) {
    this.id = id;
    this.resource = resource;
    this.capacityKg = capacityKg;
    this.amountKg = Math.min(initialKg, capacityKg);
    this.totalDepositedKg = this.amountKg;
    this.displayName = resourceDisplayName[resource];
  }

  get freeKg(): number {
    return Math.max(0, this.capacityKg - this.amountKg);
  }

  get fraction(): number {
    return this.capacityKg > 0 ? this.amountKg / this.capacityKg : 0;
  }

  /** Deposit up to kg; returns amount actually stored. Excess is vented (tracked). */
  deposit(kg: number): number {
    if (kg <= 0) return 0;
    const stored = Math.min(kg, this.freeKg);
    this.amountKg += stored;
    this.totalDepositedKg += stored;
    this.totalVentedKg += kg - stored;
    return stored;
  }

  /** Withdraw up to kg; returns amount actually obtained. */
  withdraw(kg: number): number {
    if (kg <= 0) return 0;
    const got = Math.min(kg, this.amountKg);
    this.amountKg -= got;
    this.totalWithdrawnKg += got;
    return got;
  }

  addCapacity(kg: number): void {
    this.capacityKg += kg;
  }
}

/** Registry of stores keyed by id, with per-resource aggregates. */
export class StoreSet {
  private byId = new Map<string, Store>();
  readonly all: Store[] = [];

  create(id: string, r: Resource, capacityKg: number, initialKg = 0): Store {
    if (this.byId.has(id)) throw new Error(`duplicate store id '${id}'`);
    const s = new Store(id, r, capacityKg, initialKg);
    this.byId.set(id, s);
    this.all.push(s);
    return s;
  }

  get(id: string): Store | undefined {
    return this.byId.get(id);
  }

  getOrCreate(id: string, r: Resource, capacityKg: number, initialKg = 0): Store {
    return this.byId.get(id) ?? this.create(id, r, capacityKg, initialKg);
  }

  totalAmount(r: Resource): number {
    let sum = 0;
    for (const s of this.all) if (s.resource === r) sum += s.amountKg;
    return sum;
  }
}

File: MarsHabitatSim/Assets/MarsSim/Core/Kernel/Store.cs
using System;
using System.Collections.Generic;

namespace MarsSim.Core
{
    /// <summary>Bulk commodities tracked by mass. Cabin atmosphere is NOT a store (see Habitat).</summary>
    public enum Resource
    {
        O2,             // gaseous/liquid O2 for life support reserve, kg
        N2,             // pressurant / atmosphere makeup, kg
        CO2,            // scrubbed/captured CO2 buffer for Sabatier, kg
        H2,             // hydrogen feedstock (Earth-imported or electrolysis), kg
        CH4,            // liquid methane propellant, kg
        LOX,            // liquid oxygen propellant (kept separate from breathing O2), kg
        WaterPotable,   // clean water, kg
        WaterWaste,     // grey/waste water awaiting recovery, kg
        WaterFeedstock, // mined/ISRU water not yet purified, kg
        Food,           // packaged + grown food, kg (kcal tracked alongside)
        Biomass,        // inedible plant matter, kg
        Regolith,       // excavated regolith, kg
        SparesMass,     // aggregate spares inventory, kg (units tracked by MaintenanceSystem)
    }

    public static class ResourceInfo
    {
        public static string DisplayName(Resource r) => r switch
        {
            Resource.O2 => "Oxygen reserve",
            Resource.N2 => "Nitrogen",
            Resource.CO2 => "CO₂ buffer",
            Resource.H2 => "Hydrogen",
            Resource.CH4 => "Methane (propellant)",
            Resource.LOX => "LOX (propellant)",
            Resource.WaterPotable => "Potable water",
            Resource.WaterWaste => "Wastewater",
            Resource.WaterFeedstock => "Raw water (ISRU)",
            Resource.Food => "Food",
            Resource.Biomass => "Biomass",
            Resource.Regolith => "Regolith",
            Resource.SparesMass => "Spares",
            _ => r.ToString()
        };
    }

    /// <summary>
    /// A mass buffer with capacity. All withdrawals/deposits are clamped and logged so the
    /// engine can audit conservation: sum(deposits) - sum(withdrawals) == delta(amount).
    /// </summary>
    public sealed class Store
    {
        public string Id { get; }
        public Resource Resource { get; }
        public string DisplayName { get; set; }

        public double CapacityKg { get; set; }
        public double AmountKg { get; private set; }

        /// <summary>Cumulative totals for conservation audits.</summary>
        public double TotalDepositedKg { get; private set; }
        public double TotalWithdrawnKg { get; private set; }
        public double TotalVentedKg { get; private set; }   // deposits rejected for lack of capacity

        public Store(string id, Resource resource, double capacityKg, double initialKg = 0, string displayName = null)
        {
            Id = id;
            Resource = resource;
            CapacityKg = capacityKg;
            AmountKg = Math.Min(initialKg, capacityKg);
            TotalDepositedKg = AmountKg;
            DisplayName = displayName ?? ResourceInfo.DisplayName(resource);
        }

        public double FreeKg => Math.Max(0, CapacityKg - AmountKg);
        public double Fraction => CapacityKg > 0 ? AmountKg / CapacityKg : 0;

        /// <summary>Deposit up to <paramref name="kg"/>; returns the amount actually stored. Excess is vented/discarded (tracked).</summary>
        public double Deposit(double kg)
        {
            if (kg <= 0) return 0;
            double stored = Math.Min(kg, FreeKg);
            AmountKg += stored;
            TotalDepositedKg += stored;
            TotalVentedKg += kg - stored;
            return stored;
        }

        /// <summary>Withdraw up to <paramref name="kg"/>; returns the amount actually obtained.</summary>
        public double Withdraw(double kg)
        {
            if (kg <= 0) return 0;
            double got = Math.Min(kg, AmountKg);
            AmountKg -= got;
            TotalWithdrawnKg += got;
            return got;
        }

        /// <summary>Capacity increase when new tankage/pantry lands.</summary>
        public void AddCapacity(double kg) => CapacityKg += kg;
    }

    /// <summary>Registry of stores keyed by id, with per-resource aggregate queries.</summary>
    public sealed class StoreSet
    {
        private readonly Dictionary<string, Store> _stores = new();
        private readonly List<Store> _ordered = new();

        public IReadOnlyList<Store> All => _ordered;

        public Store Create(string id, Resource r, double capacityKg, double initialKg = 0, string displayName = null)
        {
            if (_stores.ContainsKey(id)) throw new InvalidOperationException($"duplicate store id '{id}'");
            var s = new Store(id, r, capacityKg, initialKg, displayName);
            _stores[id] = s;
            _ordered.Add(s);
            return s;
        }

        public Store Get(string id) => _stores.TryGetValue(id, out var s) ? s : null;

        public Store GetOrCreate(string id, Resource r, double capacityKg, double initialKg = 0)
            => Get(id) ?? Create(id, r, capacityKg, initialKg);

        public double TotalAmount(Resource r)
        {
            double sum = 0;
            foreach (var s in _ordered) if (s.Resource == r) sum += s.AmountKg;
            return sum;
        }

        public double TotalCapacity(Resource r)
        {
            double sum = 0;
            foreach (var s in _ordered) if (s.Resource == r) sum += s.CapacityKg;
            return sum;
        }
    }
}
